pydyad/pydyad/context.py

```diff
@@ -0,0 +1,41 @@
+from pydyad.bindings import Dyad
+
+from contextlib import contextmanager
+import io
+from pathlib import Path
+
+
+DYAD_IO = None
+
+
+# The design of dyad_open is based on the PEP for Python's 'with' syntax:
+# https://peps.python.org/pep-0343/
+@contextmanager
+def dyad_open(*args, dyad_ctx=None, register_dyad_ctx=False, **kwargs):
+    global DYAD_IO
+    local_dyad_io = dyad_ctx
+    if dyad_ctx is None:
+        if DYAD_IO is None:
+            DYAD_IO = Dyad()
+            DYAD_IO.init_env()
+        local_dyad_io = DYAD_IO
+    else:
+        if register_dyad_ctx:
+            DYAD_IO = dyad_ctx
+    fname = args[0]
+    if not isinstance(args[0], Path):
+        fname = Path(args[0])
+    fname = fname.expanduser().resolve()
+    if kwargs["mode"] in ("r", "rb", "rt"):
```

Comment:
I'm new to Dyad, could you walk me through what is happening here? I'm trying to follow the test case where you submit a job with flux, and you create a consumer and produced (different jobs but to the same path?) So are they just writing / reading data from the same paths (assuming a shared filesystem) or is there a way to use some kind of different protocol? And is flux always required? And finally, how does the io.open object to relate to local_dyad_io?

Thanks for answering my questions! I hope I can help when I better understand the high level stuff.

Reply:
Thanks for asking them and thanks for the comments!

So, in general, the control- and data-flow works like this:

Producer Side:
1. The "real" close function is called
2. Producer opens a file in write-only mode (we currently don't support append or R/W, although we do have plans for the later in #29)
3. Producer closes that file
4. DYAD checks if the file is within its "producer-managed path" (set by user using the `DYAD_PRODUCER_PATH` environment variable)
   a. If the file is **not** in the producer-managed path, nothing extra happens (besides the file close)
5. DYAD registers information about the file in the Flux KVS (currently, it's just the file path relative to the producer-managed path as key and the Flux rank as value)

Consumer Side:
1. Consumer opens a file in read-only mode
2. DYAD checks if the file is within its "consumer-managed path" (set by user using the `DYAD_CONSUMER_PATH` environment variable)
   a. If the file is **not** in the consumer-managed path, nothing extra happens (besides the file open)
3. DYAD looks for an entry in the Flux KVS corresponding to the file. If there is no entry available, DYAD will block until an entry is added. This ensures the consumer is synchronized on data production (can't run a piece of code if its input isn't available, after all)
   a. If the KVS entry says that the consumer and producer are on the same node (as indicated by the Flux rank), the rest of the DYAD-specific stuff is skipped
5. DYAD creates an RPC to invoke `dyad.fetch` on the Flux broker running on the producer's node
6. The `dyad.fetch` callback reads the file (currently all at once, but we intend to change that in the future) and transmits it back to the consumer using DYAD's Data Transport Layer, or DTL (added as part of the soon-to-be-merged PR #24). By default, the DTL will use [UCX](https://openucx.org/) for the actual communication.
7. DYAD writes the file into the consumer's storage
8. The actual "open" is called

Reply:
Because of this, DYAD can work with either local or shared storage. The behavior above is what happens when using local storage. When using shared storage, we run all the steps on the producer side, but only steps 1-3 and 7 on the consumer side.

Currently, the user has to tell us when they are using shared storage using the `DYAD_SHARED_STORAGE` environment variable. However, my project this summer has been to automate that decision making using an abstraction of the storage hierarchy that I'm calling the "storage graph". The initial prototype implementation of this is almost done. Just running into some symbol export issues that refuse to be fixed by libtool's `-export-symbols` flag.

Reply:
Oh this is so cool! So the actual file open and path ate mostly just for checking permissions or ownership of the path and the real content is in flux kvs.

I know you have a little more work to do, but I’ll start a testing setup of this branch in the flux operator soon to at least try to reproduce the test. Also, your description above was hugely helpful - if you don’t have it in a README or docs somewhere I highly recommend adding it.

And it probably follows in the future that something else other than flux kvs could be used for that layer, so we could potentially run this without flux as a dependency? I’m hugely interested in this for Kubernetes!

Reply:
Most of this functionality is implemented in DYAD's core library (i.e., `src/core` in the source code and `libdyad_core.so` in an install). However, anything tied to the APIs is implemented as part of those APIs.

This is why, in `dyad_open`, we have the two conditionals wrapping the `consume` and `produce` calls. The conditional around `consume` (starting on [line 29](https://github.com/flux-framework/dyad/blob/fabc8d5f25e02292dd9ba09564e6be1bb677def9/pydyad/pydyad/context.py#L29)) is essentially performing steps 1 and 2 for the consumer side. Similarly, the conditional around `produce` (starting on [line 38](https://github.com/flux-framework/dyad/blob/fabc8d5f25e02292dd9ba09564e6be1bb677def9/pydyad/pydyad/context.py#L38)) is performing step 4 and the mode check of step 2 for the producer side.

Reply:
Now that I'm out of info dump mode, let me actually answer the specific questions you asked :sweat_smile:

**Q: "So are they just writing / reading data from the same paths (assuming a shared filesystem) or is there a way to use some kind of different protocol?"**
**A:** That depends on the values provided to the environment variables `DYAD_PRODUCER_PATH`, `DYAD_CONSUMER_PATH`, and `DYAD_SHARED_STORAGE`. You can think of the producer/consumer paths as the directories DYAD tracks to detect production/consumption events. What actual storage resources they point to is completely up to the user. So, they can point to shared storage (e.g., Lustre) or local storage (e.g., tmpfs, node-local SSD, or even El Cap Rabbit XFS). Additionally, we currently require the user to tell us if they are pointing to local or shared storage by setting `DYAD_SHARED_STORAGE` (if set, storage is shared; otherwise, storage is local). This will be automated in the near future though.

**Q: "And is flux always required?"**
**A:** Yes. Right now, DYAD uses Flux's KVS and RPC services to perform (1) information sharing and (2) control messaging for data transfer respectively. There has been some talk about looking for other options, particularly for the KVS service. But, we have no concrete plans to replace either the KVS or RPC at this time, and there are no plans to move away from Flux entirely.

**Q: "And finally, how does the io.open object to relate to local_dyad_io?"**
**A:** They aren't really related in any way. Essentially, `local_dyad_io` is used to perform the DYAD operations, and `io.open` is used to perform the "real" file I/O operations. The key detail is that `dyad_open` as a whole is essentially the glue that ensures that DYAD operations are correctly called (using `local_dyad_io`) at the correct points in the file I/O process.

Reply:
Also, thanks to writing all this out, I realized the the path checks on lines 30-31 and 39-40 are redundant because they are already performed in `libdyad_core.so`. I'll change that.

Reply:
Good catch! I’m always happy to serve as the rubber duck! 🦆 😆

Reply:
Regarding the dependency on Flux for KVS and RPC, options being considered include foundation db and grpc